### ISSUE TITLE
setOptionsCaseInsensitive should make negatable options case insensitive

### ISSUE THREAD

**Issue by remkop** [remkop/picocli#1030]
This is a follow-up ticket for #9 (case-insensitive option parsing).

As pointed out by @NewbieOrange and @ifsheldon on https://github.com/remkop/picocli/pull/1029, the negatable option transformer created by `RegexTransformer.createDefault` is always case sensitive and is not impacted by case sensitive mode.

This is not intentional: calling `setOptionsCaseInsensitive(true)`  should also make negatable options case insensitive. 

One idea is that we add a factory method `RegexTransformer::createCaseInsensitive` that would return a case insensitive transformer. Then, in the `CommandSpec::optionsCaseInsensitive` setter method, we would set the `negatableOptionTransformer` field to the case insensitive transformer. 

The javadoc for the `CommandSpec::optionsCaseInsensitive` setter and the `CommandLine::setOptionsCaseInsensitive` setter should warn users that the transformer is replaced when changing case sensitivity, so if applications have a custom transformer, they should install it _after_ changing case sensitivity.

**Comment by NewbieOrange** [comment]
I am not sure about the connection between options being case-insensitive and the transformer's case-insensitivity. For example, even if the options are case-sensitive, shouldn't --NO-VERBOSE still a valid negative form of --VERBOSE?

Probably just make the transformer case-insensitive is fine? --NO-VERBOSE is always the negative form of --VERBOSE regardless of the option's case-sensitivitiy.

**Comment by NewbieOrange** [comment]
So, ignoring the case-insensitivity part, if there is a option --NO-VERBOSE, should its negative form being --verbose or --VERBOSE? or if there is --VERBOSE, is it --NO-VERBOSE, --no-VERBOSE or something else?

I am a bit confused right now. 🤨

**Comment by NewbieOrange** [comment]
If we make the transformer case-insensitive, i.e. --VERBOSE to --no-VERBOSE and --NO-VERBOSE to --verbose, the bahaviour should make sense in case-insensitive mode, and consistent with previous behaviour (before case-insensivitivity is introduced) in case-sensitive mode.

**Comment by remkop** [comment]
Good questions. I believe it should work as follows.

## Default (case sensitive)

Option name | Negated form
------------ | -------------
`--verbose` | `--no-verbose`
`--no-verbose` | `--verbose`
`--VERBOSE` | `--no-VERBOSE`
`--no-VERBOSE` | `--VERBOSE`
`--NO-VERBOSE` | `--no-NO-VERBOSE`

Note that by default, the `--NO-VERBOSE` option is simply considered a long option, it is not considered to be the negated form of `--VERBOSE`. This is how `RegexTransformer.createDefault` is implemented and I believe this is fine.


## Case Insensitive mode
In case insensitive mode, the display label for the negated form of `--verbose` would still be `--[no-]verbose`, but the parser would recognize `--NO-VERBOSE`, `--No-Verbose`, `--nO-vErbOsE`, `--no-verbose` and more as the negated form of `--VERBOSE`, `--verbose`, `--verBOSE`, etc.

Option name | Example Match |Negated form|Example negated match
------------ | -------------|------------ | -------------
`--verbose` | `--VeRboSE`  | `--no-verbose`|`--No-VERbose`
`--no-verbose` |`--NO-VErbOsE` | `--verbose` | `--verBOSE` 
`--VERBOSE`  | `--VeRboSE` | `--no-VERBOSE`|`--NO-VERBOSe`
`--no-VERBOSE` |`--NO-VErbOsE` | `--VERBOSE`| `--VerBOSE` 
`--NO-VERBOSE` |`--NO-VErbOsE`| `--VERBOSE`| `--VERBOSe` 

The case sensitive transformer impl. could perhaps look like this (I have not tested this):

```java
CommandLine.RegexTransformer transformer = new CommandLine.RegexTransformer.Builder()
        .addPattern("(?i)^--no-(\\w(-|\\w)*)$", "--$1", "--[no-]$1")
        .addPattern("^--(\\w(-|\\w)*)$", "--no-$1", "--[no-]$1")
        .addPattern("^(-|--)(\\w*:)\\+(\\w(-|\\w)*)$", "$1$2-$3", "$1$2(+|-)$3")
        .addPattern("^(-|--)(\\w*:)\\-(\\w(-|\\w)*)$", "$1$2+$3", "$1$2(+|-)$3")
        .build();
```

By prefixing the regular expression with `(?i)`, the match becomes case insensitive.
(Need to think a bit about reducing the duplicate code.)

**Comment by NewbieOrange** [comment]
So, in #1029 , the bahaviour is

Option name | Example Match | Negated form | Example negated match
-- | -- | -- | --
--verbose | --VeRboSE | --no-verbose | --No-VERbose
--no-verbose | --NO-VErbOsE | --verbose | --verBOSE
--VERBOSE | --VeRboSE | --no-VERBOSE | --NO-VERBOSe
--no-VERBOSE | --NO-VErbOsE | --VERBOSE | --VerBOSE
--NO-VERBOSE | --NO-VErbOsE | --NO-NO-VERBOSE | --no-no-VERBOSe

This probably makes more sense? When registering a command, the behaviour about the --NO- prefix should be consistent to avoid confusion. If --NO is not negated form in case-sensitive, why would it be negated in case-insensitive?

To put it another way, why would a command changes whether it is negated or not when the case-sensitivitity change? (the only thing changes should be whether the string matching (parsing) is case-sensitive or not) This will probably introduce more confusion.

**Comment by remkop** [comment]
## What is the current spec
In the default (case sensitive) mode,  if an application creates a negatable option `--verbose`, I believe picocli should only match `--no-verbose` as the negated form. While it is true that the user manual [section on negatable options](https://picocli.info/#_negatable_options) does not explicitly mention case sensitivity, I believe most users will assume that the `--no-` prefix  is case sensitive.

## Should we change the spec?
It is possible to change this and redefine negatable options such that the `--no-` prefix is _always_ case insensitive. That way, even in the default (case sensitive) mode, picocli would match `--NO-verbose` as well as `--no-verbose` (and `--No-verbose`, `--nO-verbose`) as the negated form.

While this is possible, I don't see the benefits of doing so. I think this is more confusing, not less.
Another (minor) disadvantage is that this would mean it is no longer possible for applications to define separate `--NO-x` and `--No-x` options if the `--x` option is negatable.

If an application has all uppercase options (like `--VERBOSE`), and they want to use the `--NO-` prefix to make `--NO-VERBOSE` the negated form, then they can register a custom `INegatableOptionTransformer` to accomplish this.

**Comment by NewbieOrange** [comment]
That does make sense. How about just leave the default transformer as-is? (even if in case-insensitive still does not take --NO)

**Comment by remkop** [comment]
I believe that it is reasonable for users to expect that after setting `CommandLine.setOptionsCaseInsensitive` to `true`, negated options would also become case insensitive. To me, this would be the most "natural" behaviour.

**Comment by NewbieOrange** [comment]
Well, the negated options technically are case-insensitive, only when adding an option, the developers need to use -no instead of -NO.

My concern about the transformer behaving differently with the case-sensitive setting is, if an option is added with case-insensitive off, but then turn on case-insensitive (or vice versa), what is the expected behaviour (or if we only change the transformer, what will the current implementation behaves)?

**Comment by remkop** [comment]
> Well, the negated options technically are case-insensitive, only when adding an option, the developers need to use -no instead of -NO.

I propose that when case-insensitive mode is switched **on**, the transformer is replaced by one that uses the `"(?i)^--no-(\\w(-|\\w)*)$"` pattern so that it matches the `--no-` prefix in a case-**insensitive** manner. See the example code at the bottom of [this comment](https://github.com/remkop/picocli/issues/1030#issuecomment-625030143).

> if an option is added with case-insensitive off, but then turn on case-insensitive (or vice versa), what is the expected behaviour

Not sure I understand the question... The expected behaviour is that picocli will match option names in the user input in a case-insensitive manner.

> (or if we only change the transformer, what will the current implementation behave)?

Not sure I understand the question... The current transformer (the one returned by `RegexTransformer.createDefault` is case sensitive and this is the one used in default (case sensitive) mode.
When an application turns on case insensitive mode, I propose that this transformer is replaced by one that uses the `"(?i)^--no-(\\w(-|\\w)*)$"` pattern so that it matches the `--no-` prefix in a case-**insensitive** manner.

So, in `CommandSpec`:

```java
public CommandSpec optionsCaseInsensitive(boolean caseInsensitiveOptions) {
    if (optionsByNameMap.isCaseInsensitive() == caseInsensitiveOptions) { return this; } // no change, no action
    optionsByNameMap.setCaseInsensitive(caseInsensitiveOptions);
    negatedOptionsByNameMap.setCaseInsensitive(caseInsensitiveOptions);
    posixOptionsByKeyMap.setCaseInsensitive(caseInsensitiveOptions);

    // I propose to add this
    INegatableOptionTransformer old = negatableOptionTransformer;
    negatableOptionTransformer = caseInsensitiveOptions
            ? RegexTransformer.createCaseInsensitive()
            : RegexTransformer.createDefault();
    new Tracer().debug("optionsCaseInsensitive was set to %s. Replaced %s with %s.%n",
            caseInsensitiveOptions, old, negatableOptionTransformer);
    return this;
}
```

**Comment by NewbieOrange** [comment]
For example:
```java
class NegOptDupli {
    @Option(names = "--verbose", negatable = true) boolean verbose;
    @Option(names = "--NO-verbose") boolean noVerbose;
}
new CommandLine(new NegOptDupli()).setCaseInsensitive(true);
```
When the CommandLine is constructed, `--NO-verbose` is not a negated command, but after `setCaseInsensitive`, it should be a negated command. So we should trigger a 'reload' of all commands (adding to maps, checking duplciations) after each case-sensitivity change?

**Comment by remkop** [comment]
Aha! Now I understand. Yes, good point.

Indeed, we should revalidate duplicate options and subcommands when case sensitivity is changed.

I believe your PR https://github.com/remkop/picocli/pull/1029 may already do this, or only partially?

**Comment by NewbieOrange** [comment]
PR #1029 only checks duplications from keys in both optionMap and negatedOptionMap, that is, not calling the transformer.makeNegative() method.

To validate the maps after chaging the transformer, would probably be slightly more complicated. Since the result from `makeNagative` is possibly different, contents from `negatedOptionsByNameMap` might be outdated, basically needs to redo the `addOptionNegative` functions for all negatable options. Then, after `negatedOptionsByNameMap` is updated, there might be non-negatable options being a duplication to the 'new' negated options.

In conclusion, to make the change in transformer work, all registered options need to be somewhat 're-added'.

**Comment by remkop** [comment]
Yes, that makes sense. Thank you for pointing that out!

**Comment by NewbieOrange** [comment]
Out of interest, why do we need the `negatedOptionsByNameMap` instead of putting both normal and negated form into the `optionsByNameMap`? (besides compatibility reasons)

**Comment by remkop** [comment]
The fact that `negatedOptionsByNameMap` is a separate map makes it easier to re-register the options when the transformer is changed. 

FYI, it looks like a utility method to do this already exists: `CommandSpec::resetNegativeOptionNames`, which is called from the `CommandSpec::negatableOptionTransformer` setter.

**Comment by NewbieOrange** [comment]
Oh, I get it now. But I think without it is not harder when changing transformer? For example, picocli could remove all negated options form using the original transformer from `optionsByNameMap` then re-adds them with the new transformer, which is, not more complicated than the current 'clear then readd' method. It also comes with the benefit of less duplicate checking for duplication.
Already updated my branch with the new case-insensitive RegexTransformer and some fixes and tests. Just my two cents, but probably something can be done in 5.0.

For now, I have updated my PR with following changes:
```
* Add a new case-insensitive RegexTransformer
* Replace negatableOptionTransformer on toggle case-sensitivity
* Validate negatedOptionsByNameMap after transformer changed
* Fix null name in addOptionNegative exception message
```
Could not figure out how to name the new tests, using numbers for now.

**Comment by remkop** [comment]
Fantastic!
It is getting late for me; I will review the PR tomorrow. Many thanks!!

**Comment by remkop** [comment]
PR merged into master. 
Many thanks!!
